cpp/src/arrow/util/hashing.h

```diff
@@ -473,7 +473,9 @@ class SmallScalarMemoTable {
 
   // Copy values starting from index `start` into `out_data`
   void CopyValues(int32_t start, Scalar* out_data) const {
-    memcpy(out_data, &index_to_value_[start], size() - start);
+    if (index_to_value_.size() > 0) {
```

Comment:
I'd put a DCHECK_GE(start, 0)` just to make sure it's positive.

Reply:
Also put a ASSERT_LE in and changed to std::copy which i think fixes another bug.